### DISCUSSION TITLE
Kotlin DSL for graphql client

### DISCUSSION THREAD

**Discussion by Vano2776** [Netflix/dgs-framework#1044] [Ideas]
Why not use Kotlin DSL for build client requests ? It would be handy for building projections.
I little rewrote abstract classes query and projection, and changed serialization consider that new structure of child classes.
This is an example of updated queries compared to the old one.
Example schema:
```
type Query {
    shows(titleFilter: String): [Show]
}

type Show {
    id: ID
    title(language: String): String
    releaseYear: Int
    actors: Person
}

type Person {
    id: ID
    firstName: String
    lastName: String
}
```

Old Java DSL request:
```
val query = GraphQLQueryRequest(
            ShowsGraphQLQuery
                .newRequest()
                .titleFilter("AnyShow")
                .build(),
            ShowsProjectionRoot()
                .id()
                .title("eng")
                .parent
                .actors()
                .id()
                .firstName()
                .lastName())
            .serialize()
```

New Kotlin DSL request (with Kotlin features):
```
val queryDSL = ShowsDSLQuery("AnyShow") {
            id
            title("eng")
            actors {
                id
                firstName
                lastName
            }
        }.serialize()
```

**Comment by srinivasankavitha** [comment]
Actually, we have a PR leveraging  Kotlin DSL in codegen almost ready: https://github.com/Netflix/dgs-codegen/pull/359

**Reply by foo4u** [reply]
I see this PR was merged but the Kotlin DSL doesn't seem to be available still. @srinivasankavitha

**Comment by lthoulon-locala** [comment]
Hello there. Can you maybe point to some part of the documentation that would explain how this works ? I'm not sure how I could use a DSL approach here ?
Thanks

**Reply by Vano2776** [reply]
Set configuration property `generateKotlinClosureProjection` is true. After generated GraphQLProjection begin query building with the method call `DgsClient.buildQuery {  "dsl request body" }`.

**Reply by lthoulon-locala** [reply]
I tried this but it seems I don't have access to DgsClient. is it supposed to be generated ? I only have my types, DgsConstants and Generated.
<img width="162" alt="image" src="https://user-images.githubusercontent.com/39990276/229720014-ccbdedbd-79c2-4c83-aa65-148baf0e8441.png">

This s my configuration using maven:
```
                <configuration>
                    <schemaPaths>
                        <schemaPath>src/main/resources/schema.graphql</schemaPath>
                    </schemaPaths>
                    <language>kotlin</language>
                    <packageName>my.client.kotlin</packageName>
                    <generateClientApiV2>true</generateClientApiV2>
                    <addGeneratedAnnotation>true</addGeneratedAnnotation>
                    <!-- Needed to deserialize responses for which you haven't asked all non-optional fields -->
                    <kotlinAllFieldsOptional>true</kotlinAllFieldsOptional>
                    <!-- This allows us to use a DSL approach -->
                    <generateKotlinClosureProjections>true</generateKotlinClosureProjections>
                    <typeMapping>
                        <BigDecimal>java.math.BigDecimal</BigDecimal>
                    </typeMapping>
                </configuration>
```

with
```
        <dgs.version>6.0.1</dgs.version>
        <dgs.codegen.version>5.1.17</dgs.codegen.version>
        <graphqlcodegen-maven-plugin.version>1.30</graphqlcodegen-maven-plugin.version>
```

Is it normal that `graphql-dgs-codegen-client-core`  stayed in 5.1.17 where everything else moved up to 5.7.0 ?

**Reply by Vano2776** [reply]
I don't know the minimum version with KotlinDSL support, i'm use 5.4.0, but if you use old version maybe need replace
`<generateClientApiV2>true</generateClientApiV2>` on` <generateClient>true</generateClient>`.

**Reply by lthoulon-locala** [reply]
I'm using version 5.7.0 that is behind version 1.30 of the [graphqlcodegen-maven-plugin](https://github.com/deweyjose/graphqlcodegen). But there is that graphql-dgs-codegen-client-core that is till in version 5.1.17 and I'm not very sure what this is about. There is no greater version. Maybe @srinivasankavitha  can help ?

**Reply by mbossenbroek** [reply]
Hi all, I can help you get this working... I wrote this kotlin version of the code-gen sometime back but it's just been working nicely so I'll have to dust off my memory a bit...

The gradle config I have set to enable this are the following:
```
    language = 'kotlin'
    generateClient = true
    generateKotlinNullableClasses = true
    generateKotlinClosureProjections = true
```

In order, they do the following:
* Use kotlin instead of java
* Enable client generation at all
* Nullable classes can be used independently as the return types for queries. These classes know the difference between "I asked for this and got an explicit null response" and "I didn't ask for this and that's why it's null so I shouldn't use it". Said another way, you can construct partial response objects without having to mark everything as nullable because you might not ask for it.
* This is the bit you're asking about above, and iirc it requires the nullable objects to work. This will generate the projection classes that enable that kotlin closure style of query

The gradle plugin does also add a reference to the `graphql-dgs-codegen-shared-core` jar, which has some shared code in it. Not sure if you'll need to add that dependency manually if using maven.  The `DgsClient` class should be part of the generated code & will look something like this. It will be the main entrypoint for using this to generate queries in this way:

```
package com.netflix.project.generated

import com.netflix.graphql.dgs.codegen.GraphQLProjection
import com.netflix.project.generated.client.MutationProjection
import com.netflix.project.generated.client.QueryProjection
import com.netflix.project.generated.client.SubscriptionProjection
import graphql.language.OperationDefinition
import kotlin.String

public object DgsClient {
  public fun buildQuery(_projection: QueryProjection.() -> QueryProjection): String =
      GraphQLProjection.asQuery(OperationDefinition.Operation.QUERY, QueryProjection(), _projection)

  public fun buildMutation(_projection: MutationProjection.() -> MutationProjection): String =
      GraphQLProjection.asQuery(OperationDefinition.Operation.MUTATION, MutationProjection(),
      _projection)

  public fun buildSubscription(_projection: SubscriptionProjection.() -> SubscriptionProjection):
      String = GraphQLProjection.asQuery(OperationDefinition.Operation.SUBSCRIPTION,
      SubscriptionProjection(), _projection)
}
```

I don't believe you'll want any of `generateClientApiV2`, `addGeneratedAnnotation`, or `kotlinAllFieldsOptional`, though they may just be ignored if you switch on the two flags I mentioned above.

As to dependency versions, a project I'm using this in has the following:
```
        "com.netflix.graphql.dgs.codegen:graphql-dgs-codegen-shared-core": {
            "locked": "5.7.0"
        },
```

With the gradle buildscript having these dependencies:
```
+--- com.netflix.graphql.dgs.codegen:graphql-dgs-codegen-gradle:latest.release -> 5.7.1
|    +--- com.netflix.graphql.dgs.codegen:graphql-dgs-codegen-core:5.7.1
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.7.0 -> 1.8.10 (*)
|    |    +--- com.netflix.graphql.dgs.codegen:graphql-dgs-codegen-shared-core:5.7.1
|    |    |    +--- com.graphql-java:graphql-java:19.2
|    |    |    |    +--- com.graphql-java:java-dataloader:3.2.0
|    |    |    |    |    \--- org.slf4j:slf4j-api:1.7.30 -> 1.7.36
|    |    |    |    +--- org.reactivestreams:reactive-streams:1.0.3
|    |    |    |    \--- org.slf4j:slf4j-api:1.7.35 -> 1.7.36
|    |    |    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.7.0 -> 1.8.10 (*)
|    |    |    \--- org.jetbrains.kotlin:kotlin-reflect:1.7.0 -> 1.8.0 (*)
|    |    +--- com.netflix.graphql.dgs:graphql-dgs:5.5.1
|    |    +--- com.graphql-java:graphql-java:19.2 (*)
|    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.13.4 -> 2.14.1
|    |    +--- com.fasterxml.jackson.core:jackson-databind:2.13.4 -> 2.14.1 (*)
|    |    +--- org.slf4j:slf4j-api:1.7.35 -> 1.7.36
|    |    +--- com.squareup:javapoet:1.13.0
|    |    +--- com.squareup:kotlinpoet:1.12.0
|    |    |    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.7.0 -> 1.8.10 (*)
|    |    |    \--- org.jetbrains.kotlin:kotlin-reflect:1.7.0 -> 1.8.0 (*)
|    |    \--- com.github.ajalt.clikt:clikt-jvm:3.5.0
|    |         +--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.6.20 -> 1.8.10 (*)
|    |         \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.6.20 -> 1.8.10
|    \--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.7.0 -> 1.8.10 (*)
```

IIRC, the `graphql-dgs-codegen-client-core` was replaced by the `graphql-dgs-codegen-shared-core` mentioned above. I don't think you want that one anymore.

Hope that helps - lmk where you end up or if you need more pointers

-Matt

**Reply by lthoulon-locala** [reply]
Thanks Matt, that actually helped a lot and everything is now working like a charm. We didn't know the proper parameters to combine. This probably should be documented. Thanks again.

**Reply by mbossenbroek** [reply]
That's great to hear! I'll get those parameters added to the docs

**Comment by srinivasankavitha** [comment]
@mbossenbroek  - when you get a chance could
you please add a quick description of the settings needed to use the v2
kotlin API? Thanks!

**Comment by srinivasankavitha** [comment]
Hi - we don't officially support the maven plugin, just the gradle version.
That is maintained by the community, so would suggest reporting there for
any issues with the maven plugin.

**Reply by lthoulon-locala** [reply]
Thanks. I understand this. I actually reported some other issues there because they were linked to the maven version of the plugin itself and a colleague of mine actually did a PR there to fix those issues.
But the issue I was mentioning here was really related to which options should I be setting to activate the DSL Client. I'll be trying out what @mbossenbroek told us as really the maven plugin is just encapsulating the codegen core like the gradle plugin is. So there should not be any difference in how the generation works unless we give the wrong parameters.
So basically as you just said in your previous message, maybe just a bit of documentation would suffice ☺️

**Reply by srinivasankavitha** [reply]
Thanks so much [@Matt Bossenbroek](mailto:mbossenbroek@netflix.com) .

@lthoulon-locala - we had initially introduced this as a new, experimental API and hence deferred documenting this until we were ready to officially support it. We have a handful of users of this API regardless, so I agree, it is time to document and make it available in general.
